Improve std.typecons.Unique

mrkline opened the pull request:
Whenever D is brought up to the general programming public, the garbage collector is quickly raised as a point of contention. Regardless of how legitimate or well-informed these concerns are or are not, it would be a massive public relations boon — and great for the language, to boot — if we could trot out a solid set of RAII-based smart pointers for those who prefer to use them. We have a solid start in `std.typecons.Unique` and `std.typecons.RefCounted`. Unfortunately, these seem to be victims of bit rot and compiler bugs of days long gone.

This is the first of several pull requests that attempt to clean these up. An overview of the changes in this request is as follows (updated 2015-04-21):

- `Unique` uses `malloc` and `free` instead of the GC for backing memory storage. Unfortunately this rules out nested types for the time being (as `emplace` cannot set the frame pointer for closures).

- `std.algorithm.move` is used instead of a special `release` member function. Whether by design or by happy accident, `move` transfers ownership between `Unique` pointers in a very similar manner to C++'s `std::move` with `std::unique_ptr`. Along with being a familiar paradigm to C++ users, using `move` to transfer ownership makes more intuitive sense and builds consistency with the rest of Phobos.

- With `std.algorithm.move` transferring ownership, `release` is deprecated.

- `Unique.create` has transformed into a freestanding `unique` function. Regardless of whether or not there is language support for checking uniqueness, a utility function that creates a `Unique`, taking the same arguments as the underlying type's constructor, is extremely useful, as demonstrated by the addition of `make_unique` to C++14.

- Constructors taking a pointer have been removed, since we now control allocation ourselves.

- A new method, `get`, returns the underlying pointer, for use in functions and code that do not play a role in the life cycle of the object. Smart pointers are as much about ownership semantics as they are about allocating and freeing memory, and non-owning code should continue to refer to data using a raw pointer or a reference.

This pull request is not meant to be merged _prima facie_, but to initiate a dialogue. I strongly believe this is a good step in the right direction and would love to hear commentary from the core devs. Perhaps some of the implementation details need to be hashed out some more, but we can all agree that a stronger showing from the smart pointers in `std.typecons` will only bolster D and Phobos.

mrkline commented on std/typecons.d:
```diff
@@ -80,40 +79,69 @@ public:
+            immutable size_t allocSize = T.sizeof;
+
+        void* rawMemory = enforce(malloc(allocSize), "malloc returned null");
+        u._p = cast(RefT)rawMemory;
```
I appear to be having a segfault on classes that have a reference to their context, such as the one on the third unit test, on line 281. When I step through the constructor in GDB, `rawMemory` is correctly set to our newly allocated memory, but after stepping past this line, `u._p` is some impossibly small address, such as 0x14. It's like the assignment is ignored or the cast messes with the address somehow. Is this some failing of my understanding? The D site says that "Casting a pointer type to and from a class type is done as a type paint (i.e. a reinterpret cast)." so I'm not sure how this is happening.

mrkline replied:
```
Breakpoint 1, std.typecons.Unique!(std.typecons.__unittestL276_3().C).Unique.create!().create() (
    __HID19=0x7fffffffe0d8) at std/typecons.d:101
101	        u._p = cast(RefT)rawMemory;
(gdb) print rawMemory
$1 = (void *) 0x92d860
(gdb) n
111	            void[] init = typeid(T).init[];
(gdb) print u._p
$2 = (struct std.typecons.__unittestL276_3.C *) 0x7fffffffe060
```

ಠ_ಠ

mrkline commented:
My rationale behind doing so was that if we remove constructors accepting `RefT` and promote `create` as the way to make a `Unique`, we can completely control the life cycle of the underlying object `T`. This is the *entire* point of `Unique`, is it not? And if we completely control the life cycle, then there is no reason to rely on the GC. If you think that's too ambitious to start with or don't agree with the idea, I can certainly shelve it for now.

I would agree that making `Unique` into something that can act as a general scope guard and take arbitrary allocation and freeing functions like `std::unique_ptr` may be useful, though it seems less needed than it is in C++ given that we have `scope (exit)`.

JakobOvrum commented on std/typecons.d:
```diff
@@ -80,40 +79,69 @@ public:
+        else
+            immutable size_t allocSize = T.sizeof;
+
+        void* rawMemory = enforce(malloc(allocSize), "malloc returned null");
```
`enforce` wants to allocate to throw an exception, so you can't use it with `malloc`.

Do:
```d
void* rawMemory = malloc(allocSize);
if(!rawMemory)
    onOutOfMemoryError();
```
(See also #3031)

mrkline replied:
Thanks! For now I'll emulate what you do in #3031, and once both are hopefully merged, we can move over to that. Hopefully this will also fix the segfault I mentioned above.

mrkline replied:
@JakobOvrum After a bit of review, I recall that I was following a pattern noticed elsewhere in Phobos:

```
~/s/d/phobos [v2.067.0 ?] % ag 'enforce\(malloc'
std/container/array.d
485:            auto p = enforce(malloc(sz));

std/regex/package.d
565:        _memory = (enforce(malloc(size))[0..size]);
641:            _memory = (enforce(malloc(size))[0..size]);
671:    void[] memory = enforce(malloc(size))[0..size];

std/stdio.d
346:        _p = cast(Impl*) enforce(malloc(Impl.sizeof), "Out of memory");

std/typecons.d
4053:            _store = cast(Impl*) enforce(malloc(Impl.sizeof));

std/uni.d
1745:        auto ptr = cast(T*)enforce(malloc(T.sizeof*size), "out of memory on C heap");
6875:            auto p = cast(ubyte*)enforce(malloc(raw_cap));
6917:        ubyte* p = cast(ubyte*)enforce(malloc(3*(grow+1)));
```

Should these be corrected in another PR?

JakobOvrum replied:
#3031 fixes a few of them but focuses on more important changes, but yes, we'll fix all of them eventually.

mrkline commented:
The segfaults I've been getting seems to be caused by a separate issue in `std.conv.emplace`. I opened a bug here: https://issues.dlang.org/show_bug.cgi?id=14402

mrkline commented:
Alright. So, I've done a bit of exploring and realized that context-aware objects cannot be `emplace`d _or_ created with `new` outside their context. Because of this, I've switched back to using the GC and added the constructors that take a pointer from `new` (since `new` seems to be the only valid way to currently create a context-aware object). This PR should now pass unit tests.

So we're all on the same page, this PR now makes the following changes:

- `std.algorithm.move` becomes the canonical way to transfer ownership of a resource from one `Unique` to another. The docs and unit tests have been updated to reflect this.

- Given the previous point, `release` just destroys the owned object and nulls the pointer.

- `create` is no longer compiled out via `version(None)`.

- Releasing the owned object is now done with `destroy` followed by `GC.free`, as suggested by the "corrective action" listed for `delete` on the deprecated features page: http://dlang.org/deprecate.html#delete

JakobOvrum commented:
> Because of this, I've switched back to using the GC and added the constructors that take a pointer from `new` (since `new` seems to be the only valid way to currently create a context-aware object).

Instead of hiding this dangerous operation in an unassuming, anonymous constructor, we should probably formalize it in a named, greppable constructor function. Ideally it would be named `assumeUnique`... although that might be an unpopular choice due to std.exception.assumeUnique.

(See also http://acehreli.org/AliCehreli_assumptions.pdf)

mrkline commented:
@JakobOvrum - I couldn't agree more. It's just that they were there already, and I was going under the assumption that fewer breaking changes would make it more likely to get this PR accepted. I'm more than happy to remove those constructors and work them into something more... excplicit in its stated intentions.

JakobOvrum commented:
Alright, this is a breaking change but `Unique` is one of those fundamental types we need to get *right*. LGTM.

mrkline commented:
I can provide those, along with some better DDoc documentation in `typecons.d` itself. I'll implement the changes proposed by @JakobOvrum ASAP (hopefully tonight, possibly tomorrow).

mrkline commented:
There seems to be a general consensus that it is more important to do `Unique` _right_, even if that means breaking changes, so long as there is a clean migration path. With that in mind, I have made the following additional changes:

- Instead of constructors that take a `RefT`, Uniques can now be created one of two ways: via `.create` or `.fromNested`. The latter takes a pointer fresh from `new`, since that seems to be the only way at time of writing to create a reference to a nested object.

- `opDot` is replaced with `alias _p this`. A cursory Google search indicates that `opDot` is deprecated and that `alias this` is the preferred method. Like C++'s `unique_ptr`, `Unique` now enjoys  pointer-like operations (such as dereferencing).

- Consequently, `isEmpty` has been removed. Instead, just use `is null` as you would with a pointer. Since a pointer implicitly casts to a boolean value, `if (myUnique)` is also valid.

MartinNowak commented:
Why create? We don't have any create in Phobos. The common pattern is to use a normal constructor and maybe add a lowercase `unique` function for inference.

MartinNowak commented:
I just made a related patch for RefCounted, https://github.com/D-Programming-Language/phobos/pull/3171.

MartinNowak commented:
But assumeUnique does something very different, it casts to immutable. It was named after the precondition for converting something to immutable.

MartinNowak commented on std/typecons.d:
```diff
@@ -89,42 +96,43 @@ public:
+
+    To ensure uniqueness, be sure to provide the direct result of $(D new).
+
+    Example:
```
You can make this example a documented unittest.

mrkline replied:
There is a corresponding unit test - see https://github.com/mrkline/phobos/blob/better-unique/std/typecons.d#L225-L236

MartinNowak replied:
OK, but if you make it a [documented unittest](dlang.org/unittest.html#documented-unittest), you don't have to copy the code.

mrkline replied:
Oh, fantastic! I wasn't aware of this feature. Thanks much.

mrkline commented on std/typecons.d:
```diff
@@ -149,25 +157,37 @@ public:
+    /**
+    Returns the underlying $(D RefT) for use by non-owning code.
+
+    The holder of a $(D Unique!T) is the <em>owner</em> of that $(D T).
```
Thanks - I'll clean that up after work.

mrkline commented:
I used `create` because the function was already there. If you would like me to change that to a constructor, I'm happy to do so.

As @Dicebot discussed, I'm not touching `assumeUnique` with this PR. It's a pretty bad misnomer given that it doesn't return a `Unique`, but changing it would open a huge can of worms.

mrkline commented:
@MartinNowak A thought just occurred to me - `create` exists because it allows us to build a `Unique` for an underlying `T` whose constructor takes no arguments, no?

MartinNowak commented on std/typecons.d:
```diff
@@ -149,25 +157,37 @@ public:
+    Allows you to dereference the underlying $(D RefT)
+    and treat it like a $(D RefT) in other ways (such as comparing to null)
+    */
+    alias _p this;
```
Both of those functions can escape references making Unique unsafe.

mrkline replied:
You're right, they can, and perhaps I should add a comment saying as much. However, this is not inherently bad. `Unique` conveys _ownership_, and non-owning functions shouldn't pass around smart pointers if they aren't going to play a role in the underlying object's lifetime.

See [this bit (starting at the 14 minute mark or so)](https://youtu.be/xnqTKD8uD64?t=14m31s) of a recent Herb Sutter talk. Yes, it's about C++, but the exact same semantics are at play here and he discusses it extensively. He makes these points much more eloquently than I do, so I hope you take the time to watch it, but to summarize, doing this bakes the ownership semantics of calls directly into the function signatues. Observe:

```d
void consume(Unique!T u) // Takes ownership of u

void poke(ref T u) // Uses u but plays no impact on its lifetime

void opt(T* u) // u is optional; has no impact on its lifetime
```

mrkline replied:
Did you take the time to watch the linked section of the talk? D _can_ (unlike C++) do compile-time checks for safety using `@safe`, but since it can't verify life cycles at compile time (a la Rust), I don't know 

If we want to mark this as unsafe, that's fine, but I still strongly believe it's necessary. What's the alternative? Passing `ref Unique!T` around? That's problematic for the reasons listed above.

mrkline replied:
Point taken. My apologies if my tone was somewhat... irreverent.

So, what's the plan? Drop `get` and just pass `Unique`s by ref? And what should we do about `opDot`? I only moved away from it because there was no documentation for it everywhere, and several posts said this was because it was deprecated for `alias this`.

MartinNowak replied:
What Dicevot says. I know that talk btw.

MartinNowak replied:
Looking at [DIP25](http://wiki.dlang.org/DIP25) we should implement this like so.
```d
ref T get() return { return *_p; }
alias get this;
```
That should be safe.

MartinNowak commented on std/typecons.d:
```diff
@@ -149,25 +157,37 @@ public:
+            import core.memory : GC;
+
+            destroy(_p);
+            GC.free(cast(void*)_p);
```
Why would Unique require T to be a GC allocated object?

mrkline replied:
If you look at the commit history of this PR, you will see that I originally tried using `malloc` and `free`, but things fell apart when it came to non-static, nested classes and structs (ones that required a closure). `emplace`, in its current form, just doesn't work for them. See https://issues.dlang.org/show_bug.cgi?id=14402 and the comments I've already made on this topic [above](#issuecomment-89867749).

If I'm wrong and you _can_ initialize types requiring a closure using something besides `new`, please let me know, but I couldn't figure out any way to get them properly initialized with the frame pointer of their context.

MartinNowak replied:
If that's also true for RefCounted we should disable such types for now and find a proper solution.
Note that it possible though hacky to set the context pointer, the problem is more how to get the right context ptr. For nested struct you should be able to first initialize a normal value (compiler will fill the context ptr), then move that value. For nested classes you could have a separate argument.

rsw0x replied:
I think it's extremely important for Unique!T to be able to infer @nogc, or else its value diminishes greatly - so this should probably be done as malloc/free and the emplace issues should be fixed, no?

mrkline replied:
From my experimentation in picking apart `emplace`, the startup sequence would have to be:

1. Blit the allocated memory with `typeid(T).init`

2. Somehow assign the frame/context pointer

3. Call the constructor (Doing so without step 2, as `emplace` currently does, causes segfaults when a constructor tries to reach into the context via a null frame pointer.)

I'm all for that option, but how can it be done? @MartinNowak - you mention some way to set the context pointer - could you show this with a small code sample? How does one get the address of their current frame?

MartinNowak replied:
Only the compiler can get the context pointer, and it will only do so when you initialize a nested struct (the normal way, not by emplacing).
```d
unique(T, Args...)(Args args)
{
   auto tmp = T(args); // will init context pointer
   moveEmplace(storage, tmp);
}
```

MartinNowak replied:
Here is `moveEmplace` https://github.com/D-Programming-Language/phobos/pull/3171/files#diff-4e008aedb3026d4a84f58323e53bf017R4274.

mrkline replied:
One _cannot_ initialize a nested struct that way, so I guess they're out too.

```d
void foo(T, A...)(auto ref A args)
{
    import std.algorithm : move;
    auto tmp = T(args); // would init context pointer if it didn't crash
    auto other = move(tmp);
}

void main()
{
    int destroyed;
    struct Nested
    {
        ~this() { ++destroyed; }
    }
    auto f = foo!Nested();
    assert(destroyed == 1);
}
```

```
$ rdmd /tmp/nope.d
/tmp/nope.d(4): Error: cannot access frame pointer of nope.main.Nested
/tmp/nope.d(15): Error: template instance nope.foo!(Nested, ) error instantiating
```

MartinNowak replied:
Ah OK, I thought that would make the foo instantiation local, as we do for nested functions.
You could construct the value on the caller side `unique(T(args))`.

MartinNowak commented:
Making this a freestanding `unique` function would be the right move.

MartinNowak commented:
Let's spend a little more time on this, Unique was basically unusable before, so we have a chance to fix it and should do this thoroughly.

I think Unique should use the same storage as RefCounted, i.e. malloc/free and it should be constructible using `unique!T(args)`, `unique!T()`, and `unique(move(t))`.

mrkline commented:
Let's talk all the time we need. Since this is a fairly fundamental type, and because it's mostly unusable now, I think we all agree that we should take the time to get it done right.

Sorry for not getting to the cleanup and making the unit tests documented as we discussed above. With luck I should get to it tonight.

mrkline commented:
I've turned `Unique.create` into `unique` as requested by @MartinNowak, polished up the docs slightly, and updated the PR description.

MartinNowak commented:
No worries, let's try to get @WalterBright @schuetzm on board to advise what to do about the escaping problems.
FWIW we should implement them as sealed references.

MartinNowak commented on std/typecons.d:
```diff
@@ -64,67 +69,44 @@ static if (is(T:Object))
 else
     alias RefT = T*;
```
We should probably choose `scoped!T` as storage for classes, what does RefCounted do here?

rsw0x replied:
RefCounted doesn't support classes. https://issues.dlang.org/show_bug.cgi?id=14168

mrkline replied:
Exactly. We'll probably need to work that over as well. And `scoped!T` doesn't work for nested classes because it uses `emplace`, and AFAIK the only way to construct a nested type is with `new`, in that scope (see our previous conversation about this).

MartinNowak replied:
I think we'll figure out a solution for classes and nested classes as well. Let's take them out of the calculation for now.

schuetzm commented:
I think an extension of DIP25 (sealed references) is more likely to be the final solution than DIP69; your `return ref` suggestion is therefore the way to go.

Coincidentally, I dumped my thoughts on unique/isolated into the wiki yesterday:
http://wiki.dlang.org/User:Schuetzm/isolated

I came across an interesting difference between `Unique` and deadalnix' `@isolated` idea: `Unique` is not transitive, but `@isolated` needs to be, otherwise it couldn't be converted to `immutable` or `shared`. Not sure yet how to reconcile the two.

schuetzm commented:
Compilable example (DMD master), but without pointers:

```d
import std.c.stdlib;

struct Payload
{
    private ubyte* stuff;
    ref ubyte opIndex(size_t index) return { return stuff[index]; }
    this(size_t size) { stuff = cast(ubyte*) malloc(size); }
    ~this() { free(stuff); }
}

struct Unique(T) {
    private T payload;
    ref T get() return { return payload; }
}

auto unique(T, Args...)(Args args) {
    return Unique!T(T(args));
}

ref test() {
    auto owner = unique!Payload(42);
    return owner.get[10];
}
```
```bash
# dmd -c -dip25 xx.d
xx.d(23): Error: escaping reference to local variable owner
```

It doesn't really work yet when pointers are involved, because DIP25 only handles `ref`. This also isn't an example of transitivity; it only comes into play when you have a double reference, like `int**` or `ref int*`. My proposal handles those conservatively: on reading, the lifetime is assume to be that of the outer reference, on writing, it's assumed to be infinite.

mrkline commented:
Alright, so current suggestions are to

- Implement something similar to `get` using DIP25

- Go back to `malloc`/`free`, and deal with nested classes later.

Sorry for the delays in implementing things. My day job's been a bit hectic this week.

mrkline commented:
So, those two are suggestions are implemented. Points of discussion:

1. As mentioned [above](https://github.com/D-Programming-Language/phobos/pull/3139#discussion_r28573873), nested structs and classes are now a no-go. There is no way (again, AFAIK, please prove me wrong) to get the frame pointer for a nested type into `unique()`.

2. Since DIP25 handles `ref` types, this seems extremely clumsy with regards to class types, as `get()` now returns a reference to a reference (since class types are reference types).

Is the current state of `get()` what you guys had in mind?

ntrel commented:
Just saw this, using malloc/free is great work.

> release just destroys the owned object and nulls the pointer

This seems like unnecessary breakage, or is there a reason for this? Why not just deprecate `release`? Adding the new `release` seems to be just a duplicate of `destroy(myUnique)`.

> this seems extremely clumsy with regards to class types, as get() now returns a reference to a reference 

I think `Unique!Object` types shouldn't have `get` return by `ref`, and there currently isn't a way to stop the class reference escaping, we need `scope` e.g. DIP69.

MartinNowak commented:
> Since DIP25 handles ref types, this seems extremely clumsy with regards to class types, as get() now returns a reference to a reference (since class types are reference types).

Yes, that doesn't make sense, it would allow you to set `_p` to something different than the `malloc`'ed memory. You could wrap classes into `scoped` and return a ref to that, but `scoped` isn't yet safe.
I think we should disallow classes until a good/common solution has been found for `Unique` and `RefCounted`.

MartinNowak commented:
> This seems like unnecessary breakage, or is there a reason for this? Why not just deprecate release? Adding the new release seems to be just a duplicate of destroy(myUnique).

Ideally we could implement `release` to release the ownership of the underlying object (unsafe) and return it. But I don't see how one could reliably destroy a released value, because it requires to know the implementation of `Unique`, i.e. `free` and `GC.removeRange`.

rsw0x commented:
re:classes,
Someone should probably ask Walter or Andrei what the status of DIP74 is, as that seems to directly relate to this/overlap(?).

ntrel commented:
> I think we should disallow classes until a good/common solution has been found

Personally I would allow classes (without `ref`) but note in the docs that scope enforcement will be added later. If we stop supporting classes now there would be no upgrade path for *correct, safe* code already using `Unique!Class`. We can't know how much code uses that, but it seems an important principle to not break working code without providing an equivalent alternative. (Breaking `Unique!T(new T)` is justifiable breakage though to allow `@nogc`).

mrkline commented:
1. You're right, `destroy` can do the job of `release`, and I should just put it back to what it did originally and deprecate it.

2. I'm strongly with @ntrel here. Saying "sorry, classes aren't supported yet until some unspecified time in the future when we offer better safety" is a bad idea. We should focus on getting the interface correct now and adding safety as the language evolves to support it. That way, there won't be changes people need to migrate to in the future, and users who are already using it safely will, without code changes, start getting additional safety as it is introduced to the language.

3. As everyone else has said, I would be incredibly curious as to what @andralex and @WalterBright have to say about this.

MartinNowak commented:
Then it should be `T get() @system` for classes, until we can control escaping.

MartinNowak commented on std/typecons.d:
```diff
@@ -178,96 +177,200 @@ private:
+        import std.algorithm : move;
+        auto temp = T(args);
+        u._p = cast(T*)rawMemory;
+        *u._p = move(temp);
```
That doesn't work like that, because you're assigning to uninitialized memory, see [RefCounted.move](https://github.com/D-Programming-Language/phobos/blob/f7498ad8ec147114b1bcbc61d0e4db3e033417f5/std/typecons.d#L4552) for how to do `moveEmplace`.

mrkline replied:
I'll go back to just using `emplace` as I was originally. There's no benefit from creating a struct and then moving it in as opposed to just calling `emplace` AFAIK.

MartinNowak commented:
We could make `Unique` and `RefCounted` use the same internal storage struct, to encapsulate common construct/move/memory code.

JakobOvrum commented:
It's not an issue of classes vs anything else, it's an issue of types with [indirection](http://dlang.org/phobos/std_traits.html#hasIndirections).

JakobOvrum commented:
@Dicebot, right, I don't think I've wrapped my head around DIP25 completely yet.

---

Regarding C heap vs GC heap; if we want to allow transferring `Unique!T` (where `T` is unshared and mutable/const)  to other threads, using the C heap might be a big advantage, as there are GC improvement ideas floating around involving thread-local GC heaps (@deadalnix?). It's best if we can decide the sharedness of a GC-allocated chunk at allocation-time.

JakobOvrum commented:
Another observation regarding cross-thread moves - the uniqueness of `Unique!T` only applies to `T`, not references embedded as `T`'s fields. That is, the uniqueness is not transitive, it only applies for one level of indirection - the `Unique!T` reference itself (as has been observed by others before me). 

That means not just any `Unique` instance can be moved to a different thread, only instances where all of `T`'s fields passes `isUnique!U || !hasUnsharedAliasing!U`, where `U` is the type of the field.

Does this make sense?

MartinNowak commented:
Can we please keep that highly speculative GC discussion out of here?
The idea of using shared to separate thread-local from global allocations is pretty old and has many pitfalls.

> That means not just any Unique instance can be moved to a different thread, only instances where all of T's fields passes isUnique!U || !hasUnsharedAliasing!U, where U is the type of the field.

Makes sense, and once we merge this we should make std.concurrent aware of Unique.

MartinNowak commented on std/typecons.d:
```diff
@@ -64,67 +69,18 @@ static if (is(T:Object))
-    Unique!C uc = new C;
-    Unique!Object uo = uc.release;
+    Unique!C uc = unique!C();
+    Unique!Object uo = move(uc);
```
Don't think that polymorph conversion works, should turn this into a documented unittest.

MartinNowak replied:
Oh, it's actually is supposed to work.

MartinNowak commented on std/typecons.d:
```diff
@@ -148,26 +104,72 @@ public:
+    bool opCast(T : bool)() const { return !empty; }
+
+    /**
+    Allows you to reference the underlying $(D RefT)
```
Forwards to the underlying value.

MartinNowak commented on std/typecons.d:
```diff
@@ -148,26 +104,72 @@ public:
+        return _p;
+    }
+
+    @property bool empty() const
```
Missing docs.

MartinNowak commented on std/typecons.d:
```diff
@@ -148,26 +104,72 @@ public:
+        return _p is null;
+    }
+
+    bool opCast(T : bool)() const { return !empty; }
```
Docs

MartinNowak commented on std/typecons.d:
```diff
@@ -178,96 +180,198 @@ private:
+    args = Arguments to pass to $(D T)'s constructor.
+
+*/
+Unique!T unique(T, A...)(auto ref A args)
```
We should also add a move constructor for `unique!T(T())`, that is when you move a complete T value into unique.

MartinNowak commented:
Looks pretty good already.

mrkline commented:
Alright, I added the suggested docs and made `get` a template function (seemed like the simplest solution). I agree that a move constructor for a `T` rvalue would be nice, but perhaps we can save that for another PR that also lifts your `moveEmplace` functionality from #3171 into its own function.

MartinNowak commented:
Auto-merge toggled on

MartinNowak commented:
Thanks, I'll follow up with RefCounted support for classes and factoring out `moveEmplace`.
